lisa/core/node.py

```diff
@@ -35,17 +41,55 @@ def createNode(
 
     def setConnectionInfo(
         self,
-        address: str = "",
-        port: int = 22,
-        publicAddress: str = "",
-        publicPort: int = 22,
-        username: str = "root",
-        password: str = "",
-        privateKeyFile: str = "",
-    ):
-        self.connection = SshConnection(
-            address, port, publicAddress, publicPort, username, password, privateKeyFile
-        )
+        address: Optional[object],
+        port: Optional[object],
+        publicAddress: Optional[object],
+        publicPort: Optional[object],
+        username: Optional[object],
+        password: Optional[object],
+        privateKeyFile: Optional[object],
+    ) -> None:
+        if self.connection is not None:
+            raise Exception(
+                "node is set connection information already, cannot set again"
+            )
+        parameters: Dict[str, object] = dict()
+        if address is not None:
+            parameters["address"] = address
+        if port is not None:
+            parameters["port"] = port
+        if publicAddress is not None:
+            parameters["publicAddress"] = publicAddress
+        if publicPort is not None:
+            parameters["publicPort"] = publicPort
+        if username is not None:
+            parameters["username"] = username
+        if password is not None:
+            parameters["password"] = password
+        if privateKeyFile is not None:
+            parameters["privateKeyFile"] = privateKeyFile
+        self.connection = SshConnection(**parameters)
+
+    def execute(self, cmd: str) -> ExecutableResult:
+        result: ExecutableResult
+        cmd_id = random.randint(0, 10000)
```

Comment:
UUID is too long to show in log. The id is for corelate informatons in log. As the life cycle of a command won't be long in test run, and we don't need to save them in separated data storage. The id helps people to associate different lines in log for troubleshooting, so every line in log need this value.

Reply:
That's an excellent point, UUIDs are annoying to display in a log. However, a random in between 0 and 10000 is far from guaranteed to be unique. Do we know the approximate max number of commands that may run during the lifetime of an execution of LISA? Suppose we have 100 tests, each test running 10 processes (and I think these are conservative estimates), then we have on the order of 1000 commands we need to uniquely identify. With the given range, we are practically guaranteed to have a collision [(if I did my math right)](https://www.wolframalpha.com/input/?i=1-+e%5E%28%28-1000*%281000-1%29%29%2F%282*10000%29%29). An alternative would be to start at 1000 and auto-increment from there. This would require one global state, but in this case (since we don't want to use a UUID) I think it makes sense, since we _don't_ want a collision (or logs would become quite confusing).

Reply:
Increasing number may not good on concurrency requesting. I will check if Python supports atom exchange increasing. But collision is acceptable, as cmd id is not expected to be unique in whole run, just around it's log. So, if there are 10 concurrent cmds, the chance may be 10/10000? Anyway, I will check concurrent part of Python, that's anyway needed in other scenarios.

Reply:
From some quick research, Python does have locks, and a lock around a single integer increasing it should not be an issue (there's nothing that would result in a deadlock).

Reply:
Some locks are low performance, but atomic exchange or spin locker pattern is high effecient. But another problem is, the test cases may run in parallel by spawn or forked processes. the lock may not cross on such cases. Another solution is to add more prefix like node/case, so that the number can be sequential in a unique way, but the prefix is longer. We can revisit it later.